### PR TITLE
ZIR-174: Iddo/keccak wip

### PR DESCRIPTION
**Pull request by iddo-bentov**

### CODE REVIEW

File: zirgen/circuit/keccak/README.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+# The Keccak Accelerator
+
+This directory contains the implementation of the keccac accelerator.
```
Comment (pdg744) [inline zirgen/circuit/keccak/README.md:3]:
```suggestion
This directory contains the implementation of the keccak accelerator.
```

File: zirgen/circuit/keccak/README.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+In this code snippet, `hasher` carries an internal state which is modified the input using the `update(input)` function. `finalize` will generate a resulting hash. In order to use the accelerator we need to generate an input transcript. To keep the proof under 2 million cycles the maximum amount of data we can hash in the accelerator is approximately 95 kilobytes.
+#### WIP: Non-concurrent implementation for crates such as tiny-keccak and sha3
+
+The following assumes a situation where a single call to finalize is compressing less than 95 KB of input data and only one instance of hasher is present.
+
+* Every time update is called, the input is split up to fit inside the 136 byte block. Once the block fills up, it is compressed using keccak. In our design, we send each of the blocks to the host in order to record the raw data of the input transcript.
+* Once the hasher's `finalize` function is called we can communicate with the termination so that the block count can be computed and for the keccak hash to be generated.
+
+#### WIP: Non-concurrent implementation for crates such as sha3
```
Comment (pdg744) [inline zirgen/circuit/keccak/README.md:48]: This appear to be redundant with the previous section.
```suggestion
```

File: zirgen/circuit/keccak/README.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+#### WIP: concurrent implementation
```
Comment (pdg744) [inline zirgen/circuit/keccak/README.md:50]: @iddo-bentov can you remind me what we've decided to do here? is this a WIP for the nearterm or something we've decided to punt on for now? 

might make sense to strike this section from the public readme and update our linear board to reflect the work we have planned here.

File: zirgen/circuit/keccak/README.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+// write input message
+hasher.update(&input);
+
+// read hash digest
+let result = hasher.finalize();
+```
+
+In this code snippet, `hasher` carries an internal state which is modified the input using the `update(input)` function. `finalize` will generate a resulting hash. In order to use the accelerator we need to generate an input transcript. To keep the proof under 2 million cycles the maximum amount of data we can hash in the accelerator is approximately 95 kilobytes.
+#### WIP: Non-concurrent implementation for crates such as tiny-keccak and sha3
```
Comment (pdg744) [inline zirgen/circuit/keccak/README.md:41]: What do we need in order to do in order to finish these WIP sections? 
I guess the details of this section will be clarified in the process of the work for ZIR-147?

File: zirgen/circuit/keccak/README.md (hunk 3 above)
Comment (iddo-bentov) [reply zirgen/circuit/keccak/README.md:50]: Indeed there's no need for the WIP comments to be in the public readme.
The current version is also the most efficient version nearterm, we will assess the concurrent version after we finish integrating this version.

**Pull request closed without merging**